File: rust/src/_primitives.rs
#[allow(unused)]
pub(super) fn ser_u8(value: u8) -> Vec<u8> {
  vec![value]
}

#[allow(unused)]
pub(super) fn de_u8(bytes: Vec<u8>) -> Option<u8> {
  bytes.get(0).copied()
}

#[allow(unused)]
pub(super) fn ser_u16(value: u16) -> Vec<u8> {
  value.to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_u16(bytes: Vec<u8>) -> Option<u16> {
  bytes.try_into().map(|v| u16::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_u32(value: u32) -> Vec<u8> {
  value.to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_u32(bytes: Vec<u8>) -> Option<u32> {
  bytes.try_into().map(|v| u32::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_u64(value: u64) -> Vec<u8> {
  value.to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_u64(bytes: Vec<u8>) -> Option<u64> {
  bytes.try_into().map(|v| u64::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_i8(value: i8) -> Vec<u8> {
  vec![value as u8]
}

#[allow(unused)]
pub(super) fn de_i8(bytes: Vec<u8>) -> Option<i8> {
  bytes.get(0).map(|&u| u as i8)
}

#[allow(unused)]
pub(super) fn ser_i16(value: i16) -> Vec<u8> {
  (value as u16).to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_i16(bytes: Vec<u8>) -> Option<i16> {
  bytes.try_into().map(|v| i16::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_i32(value: i32) -> Vec<u8> {
  (value as u32).to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_i32(bytes: Vec<u8>) -> Option<i32> {
  bytes.try_into().map(|v| i32::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_i64(value: i64) -> Vec<u8> {
  (value as u64).to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_i64(bytes: Vec<u8>) -> Option<i64> {
  bytes.try_into().map(|v| i64::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_f32(value: f32) -> Vec<u8> {
  value.to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_f32(bytes: Vec<u8>) -> Option<f32> {
  bytes.try_into().map(|v| f32::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_f64(value: f64) -> Vec<u8> {
  value.to_le_bytes().to_vec()
}

#[allow(unused)]
pub(super) fn de_f64(bytes: Vec<u8>) -> Option<f64> {
  bytes.try_into().map(|v| f64::from_le_bytes(v)).ok()
}

#[allow(unused)]
pub(super) fn ser_string(value: String) -> Vec<u8> {
  let mut v = value.as_bytes().to_vec();
  v.push(0);
  v
}

#[allow(unused)]
pub(super) fn de_string(mut bytes: Vec<u8>) -> Option<String> {
  String::from_utf8(bytes).ok()
}

File: ts/src/_primitives.ts
export function serU8(value: number): number[] {
  const buf = new ArrayBuffer(1);
  const dv = new DataView(buf);
  dv.setUint8(0, value);
  return [...new Uint8Array(buf)];
}
export function deU8(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getUint8(0);
}

export function serU16(value: number): number[] {
  const buf = new ArrayBuffer(2);
  const dv = new DataView(buf);
  dv.setUint16(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deU16(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getUint16(0, true);
}

export function serU32(value: number): number[] {
  const buf = new ArrayBuffer(4);
  const dv = new DataView(buf);
  dv.setUint32(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deU32(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getUint32(0, true);
}

export function serU64(value: bigint): number[] {
  const buf = new ArrayBuffer(8);
  const dv = new DataView(buf);
  dv.setBigUint64(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deU64(bytes: number[]): bigint | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getBigUint64(0, true);
}

export function serI8(value: number): number[] {
  const buf = new ArrayBuffer(1);
  const dv = new DataView(buf);
  dv.setInt8(0, value);
  return [...new Uint8Array(buf)];
}
export function deI8(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getInt8(0);
}

export function serI16(value: number): number[] {
  const buf = new ArrayBuffer(2);
  const dv = new DataView(buf);
  dv.setInt16(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deI16(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getInt16(0, true);
}

export function serI32(value: number): number[] {
  const buf = new ArrayBuffer(4);
  const dv = new DataView(buf);
  dv.setInt32(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deI32(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getInt32(0, true);
}

export function serI64(value: bigint): number[] {
  const buf = new ArrayBuffer(8);
  const dv = new DataView(buf);
  dv.setBigInt64(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deI64(bytes: number[]): bigint | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getBigInt64(0, true);
}

export function serF32(value: number): number[] {
  const buf = new ArrayBuffer(4);
  const dv = new DataView(buf);
  dv.setFloat32(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deF32(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getFloat32(0, true);
}

export function serF64(value: number): number[] {
  const buf = new ArrayBuffer(8);
  const dv = new DataView(buf);
  dv.setFloat64(0, value, true);
  return [...new Uint8Array(buf)];
}
export function deF64(bytes: number[]): number | undefined {
  const dv = new DataView(new Uint8Array(bytes).buffer);
  return dv.getFloat64(0, true);
}

export function serString(value: string): number[] {
  return [...new TextEncoder().encode(value), 0];
}
export function deString(bytes: number[]): string | undefined {
  return new TextDecoder().decode(new Uint8Array(bytes.slice(0, -1)));
}

export function serBool(value: boolean): number[] {
  return [value ? 1 : 0];
}
export function deBool(value: number[]): boolean | undefined {
  if (value[0] !== undefined) {
    return value[0] > 0;
  } else {
    return undefined;
  }
}
